docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,224 @@
+```
+
+We introduce the following two queries:
+```bash
+interchain-security-pd query provider optedInValidators $chainID
+interchain-security-pd query provider hasToValidate $providerAddr
+```
+One query to retrieve the validators that are opted in and hence the validators that need to validate the consumer chain and one query that given a validator's address returns all the chains this validator has to validate.
+
+#### When do validators opt in?
+As described earlier, validators can manually opt in by sending a `MsgOptIn` message.
+Additionally, in a Top N chain, a validator is automatically opted in when it moves from the bottom `(100 - N)%` to the top `N%` validators.
+
+Lastly, validators can also opt in if they vote `Yes` during the `ConsumerAdditionProposal` that introduces a consumer chain.
+This simplifies validators operations because they do not have to send an additional message to opt in.
+
+Because the `Tally` method [deletes the votes](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/gov/keeper/tally.go#L71) after reading them, we cannot check the votes of the validators after the votes have been tallied.
+To circumvent this, we introduce a hook for [`AfterProposalVote`](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/gov/keeper/vote.go#L35) and keep track of all the votes cast by a validator.
+If a validator casts more than one vote, we only consider the latest vote.
+Finally, we only consider a validator has opted in if it casts a 100% `Yes` vote in case of a [weighted vote](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-037-gov-split-vote.md).
+
+
+### How do validators opt out?
+Validators that have opted in on a chain can opt out by sending the following message:
+
+```protobuf
+message MsgOptOut {
+    // the chain id of the consumer chain to opt out from
+    string chainID = 1;
+    // the provider address of the validator
+    string providerAddr = 2;
+}
+```
+Validators can only opt out after a consumer chain has started and hence the above message returns an error if the chain with `chainID` is not running.
+Additionally, a validator that belongs to the top `N%` validators cannot opt out from a Top N chain and hence a `MsgOptOut` would error in such a case.
+
+#### State & Query
+We also update the state of the opted-in validators when a validator has opted out by removing the opted-out validator.
+
+Note that we only punish validators for downtime that were  opted in on a consumer chain.
+For this, we keep historical info of all the validators that have opted in during the last `X` (to be defined) blocks.
+This way we can jail validators for downtime knowing that indeed the validators have opted in at some point in the past. 
+Otherwise, we can think of a scenario where a validator `V` is down for a period of time, but before `V` gets punished for downtime, validator `V` opts out, and then we do not know whether `V` should be punished or not.
```

Comment:
We could use the opted in state `optedInBytePrefix | len(chainID) | chainID | addr` and add the blockheight as a value.

Reply:
We intend to store the blockheight as a value in any case. See above:
> For each validator, we store on whether the validator has opted in and at which block height the validator opted in under the key:
```
optedInBytePrefix | len(chainID) | chainID | addr
```
I think we can introduce one more `boolean` flag on whether the validator is currently opted in or not and have something similar to `addr -> (blockHeight, isOptedIn)` and this way if someone opts out we just set the `boolean` flag to `false`.  But we would still keep the information that the validator was opted in at some earlier point in time and at which height.

Reply:
Updated the ADR to capture the above. Resolving this.